src/main/scala/inox/solvers/theories/ASCIIStringEncoder.scala

```diff
@@ -108,7 +108,8 @@ trait ASCIIStringEncoder extends SimpleEncoder {
 }
 
 object ASCIIStringEncoder {
-  def apply(p: Program): ASCIIStringEncoder { val sourceProgram: p.type } = new {
-    val sourceProgram: p.type = p
-  } with ASCIIStringEncoder
+  def apply(p: Program): ASCIIStringEncoder { val sourceProgram: p.type } = {
```

Comment:
The local class shouldn't be needed, you should be able to directly write:
```scala
def apply(p: Program): ASCIIStringEncoder { val sourceProgram: p.type } = new ASCIIStringEncoder {
  val sourceProgram: p.type = p
}
````

Reply:
The methods used at the beginning of the `ASCIIStringEncoder`
https://github.com/epfl-lara/inox/blob/d13789665dbc9b8ce7a32557d68460d11307fa67/src/main/scala/inox/solvers/theories/ASCIIStringEncoder.scala#L17-L28 like `mkSort`, `mkFunDef`, `T`, etc. depend on
https://github.com/epfl-lara/inox/blob/d13789665dbc9b8ce7a32557d68460d11307fa67/src/main/scala/inox/solvers/theories/TheoryEncoder.scala#L11 so we need the `sourceProgram` to be initialized before the `ASCIIStringEncoder` constructor gets called. We could maybe use lazy vals for all these fields but we had a bunch of deadlocking issues due to laziness in the project so we'd prefer not having to introduce new ones.

Trait parameters would be nicer but we'd like to keep source compatibility with 2.11 for now.

Reply:
Do you still want to maintain source compatibility with 2.11 now that 2.13 is around the corner? Supporting the last two Scala version + Dotty sounds like a good deal even if that means dropping 2.11 no?